frontend/projects/ui/src/app/services/dep-error.service.ts

```diff
@@ -0,0 +1,210 @@
+import { Injectable } from '@angular/core'
+import { Emver } from '@start9labs/shared'
+import { map, shareReplay } from 'rxjs/operators'
+import { PatchDB } from 'patch-db-client'
+import {
+  DataModel,
+  HealthCheckResult,
+  HealthResult,
+  InstalledPackageDataEntry,
+  PackageMainStatus,
+} from './patch-db/data-model'
+
+export type PackageDependencyErrors = Record<string, DependencyErrors>
+export type DependencyErrors = Record<string, DependencyError | null>
+
+@Injectable({
+  providedIn: 'root',
+})
+export class DepErrorService {
+  readonly depErrors$ = this.patch.watch$('package-data').pipe(
+    map(pkgs =>
+      Object.keys(pkgs)
+        .sort((a, b) =>
+          dependencyDepth(pkgs, b) > dependencyDepth(pkgs, a) ? -1 : 1,
+        )
+        .reduce((errors, id) => {
+          const toReturn: PackageDependencyErrors = {
+            ...errors,
+            [id]: this.getDepErrors(pkgs, id, errors),
+          }
+
+          return toReturn
+        }, {} as PackageDependencyErrors),
+    ),
+    shareReplay(1),
+  )
+
+  constructor(
+    private readonly emver: Emver,
+    private readonly patch: PatchDB<DataModel>,
+  ) {}
+
+  private getDepErrors(
+    pkgs: DataModel['package-data'],
+    pkgId: string,
+    outerErrors: PackageDependencyErrors,
+  ): DependencyErrors {
+    const pkgInstalled = pkgs[pkgId].installed
+
+    if (!pkgInstalled) return {}
+
+    return currentDeps(pkgs, pkgId).reduce((innerErrors, depId) => {
+      const toReturn: Record<string, DependencyError | null> = {
+        ...innerErrors,
+        [depId]: this.getDepError(pkgs, pkgInstalled, depId, outerErrors),
+      }
+      return toReturn
+    }, {} as DependencyErrors)
+  }
+
+  private getDepError(
+    pkgs: DataModel['package-data'],
+    pkgInstalled: InstalledPackageDataEntry,
+    depId: string,
+    outerErrors: PackageDependencyErrors,
+  ): DependencyError | null {
+    const depInstalled = pkgs[depId]?.installed
```

Comment:
🤔 We could have broken this up into 4 fns, is it better?
I thought of it because you where commenting sections of code, and the built in way of naming sections of code is fns

```ts
return this.getDepErrorInstalled(pkgs, pkgInstalled, depId, outerErrors) |
 this.getDepErrorIncorrectVersion(pkgs, pkgInstalled, depId, outerErrors) |
  this.getDepErrorInvalidConfig(pkgs, pkgInstalled, depId, outerErrors) |
    this.getDepErrorNotRunning(pkgs, pkgInstalled, depId, outerErrors) |
    null
}
private getDepErrorInstalled(
    pkgs: DataModel['package-data'],
    pkgInstalled: InstalledPackageDataEntry,
    depId: string,
    outerErrors: PackageDependencyErrors,
  ): DependencyError | null {
      const depInstalled = pkgs[depId]?.installed
     if (depInstalled) return null;
     return {
        type: DependencyErrorType.NotInstalled,
      }
  }
  private getDepErrorIncorrectVersion(
    pkgs: DataModel['package-data'],
    pkgInstalled: InstalledPackageDataEntry,
    depId: string,
    outerErrors: PackageDependencyErrors,
  ): DependencyError | null {
      const depInstalled = pkgs[depId]?.installed
     if (!depInstalled) return null;
     
    const pkgManifest = pkgInstalled.manifest
    const depManifest = depInstalled.manifest
    if (
      this.emver.satisfies(
        depManifest.version,
        pkgManifest.dependencies[depId].version,
      )
    )  return null
    
      return {
        type: DependencyErrorType.IncorrectVersion,
        expected: pkgManifest.dependencies[depId].version,
        received: depManifest.version,
      }
    
  }